examples/mpt/main.cpp

```diff
@@ -0,0 +1,737 @@
+                       ggml_cpy(ctx0, Qcur,
+                                ggml_new_tensor_3d(ctx0, GGML_TYPE_F32,
+                                                   n_embd / n_head, n_head, N)),
+                       0, 2, 1, 3);
+
+      // K = Kmem.view(n_embd/n_head, n_head, n_past + N).permute(0, 2, 1, 3)
+      // [64, n_past + N, 12]
+      struct ggml_tensor *K = ggml_permute(
+          ctx0,
+          ggml_reshape_3d(
+              ctx0,
+              ggml_view_1d(ctx0, model.memory_k, (n_past + N) * n_embd,
+                           il * n_ctx * ggml_element_size(model.memory_k) *
+                               n_embd),
+              n_embd / n_head, n_head, n_past + N),
+          0, 2, 1, 3);
+      // K * Q
+      struct ggml_tensor *KQ = ggml_mul_mat(ctx0, K, Q);
+
+      // KQ_scaled = KQ / sqrt(n_embd/n_head)
+      struct ggml_tensor *KQ_scaled = ggml_scale(
+          ctx0, KQ, ggml_new_f32(ctx0, 1.0f / sqrt(float(n_embd) / n_head)));
+
+      struct ggml_tensor *KQ_scaled_alibi =
+          ggml_alibi(ctx0, KQ_scaled, n_past, n_head);
+
+      // KQ_masked = mask_past(KQ_scaled)
+      struct ggml_tensor *KQ_masked =
+          ggml_diag_mask_inf(ctx0, KQ_scaled_alibi, n_past);
+
+      // KQ = soft_max(KQ_masked)
+      struct ggml_tensor *KQ_soft_max = ggml_soft_max(ctx0, KQ_masked);
+
+      // V_trans = Vmem.view(n_embd/n_head, n_head, n_past + N).permute(1, 2,
+      // 0, 3).contiguous() [n_past + N, 64, 12]
+      struct ggml_tensor *V_trans = ggml_cpy(
+          ctx0,
+          ggml_permute(
+              ctx0,
+              ggml_reshape_3d(
+                  ctx0,
+                  ggml_view_1d(ctx0, model.memory_v, (n_past + N) * n_embd,
+                               il * n_ctx * ggml_element_size(model.memory_v) *
+                                   n_embd),
+                  n_embd / n_head, n_head, n_past + N),
+              1, 2, 0, 3),
+          ggml_new_tensor_3d(ctx0, model.memory_v->type, n_past + N,
+                             n_embd / n_head, n_head));
+
+      // KQV = transpose(V) * KQ_soft_max
+      struct ggml_tensor *KQV = ggml_mul_mat(ctx0, V_trans, KQ_soft_max);
+
+      // KQV_merged = KQV.permute(0, 2, 1, 3)
+      struct ggml_tensor *KQV_merged = ggml_permute(ctx0, KQV, 0, 2, 1, 3);
+
+      // cur = KQV_merged.contiguous().view(n_embd, N)
+      cur = ggml_cpy(ctx0, KQV_merged,
+                     ggml_new_tensor_2d(ctx0, GGML_TYPE_F32, n_embd, N));
+
+      // projection
+      {
+        cur = ggml_mul_mat(ctx0, model.layers[il].c_attn_out_proj_weight, cur);
+      }
+    }
+
+    inpL = ggml_add(ctx0, inpL, cur);
+
+    // m = self.ln_2(x)
+    {
+      cur = ggml_norm(ctx0, inpL);
+
+      cur = ggml_mul(
+          ctx0, ggml_repeat(ctx0, model.layers[il].norm_2_weight, cur), cur);
+    }
+
+    // n = self.mlp(m)
+    {
+
+      cur = ggml_mul_mat(ctx0, model.layers[il].ffn_up_proj, cur);
+
+      // GELU activation
+      cur = ggml_gelu(ctx0, cur);
```

Comment:
Shouldn't we use `ggml_relu()` here?

https://huggingface.co/mosaicml/mpt-7b/blob/main/config.json#L31